### PR TITLE
Fix table iterator error for configurables

### LINKED ISSUE #28676: Bad, sad error when iterating over a table constructed for a configurable variable at runtime
**Description:**
$title. 

Happens in the `io.ballerina.runtime.internal.TypeChecker#checkIsTableType` check at https://github.com/ballerina-platform/ballerina-lang/blob/master/bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java#L1020 because `io.ballerina.runtime.internal.TypeChecker#getTableConstraintField` returns null when the constraint type is an intersection type.

May be an issue with non-configurable table values too (if it is possible that we get here and the constraint type is an intersection type). 

**Steps to reproduce:**
```ballerina
type AuthInfo record {
    readonly string username;
    string password;
    string[] scopes?;
};

configurable table<AuthInfo> key(username) & readonly users = ?;

public function main() {
    foreach var item in users {
    }
}
```

**Affected Versions:**
slalpha2

### PR DESCRIPTION
**Pull request by HindujaB**
## Purpose
The issue happens when an Intersection type (`record & readonly`) is passed as a table constraint type when creating Table value. 
Fixes #28676 

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [x] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Comment by warunalakshitha** [comment]
LGTM

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/configurable/ConfigTomlParser.java [hunk 1]
```diff
@@ -277,7 +277,9 @@ private static Object retrieveTableValues(TomlTableArrayNode tomlValue, String v
                 new ArrayValueImpl(TypeCreator.createArrayType(constraintType), tableSize, tableEntries);
         ArrayValue keyNames = keys == null ? (ArrayValue) ValueCreator.createArrayValue(new BString[]{}) :
                 (ArrayValue) StringUtils.fromStringArray(keys);
-        return new TableValueImpl<>((BTableType) tableType, tableData, keyNames);
+        TableType type =
+                TypeCreator.createTableType(((IntersectionType) constraintType).getEffectiveType(), keys, true);
```
Comment (HindujaB) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/configurable/ConfigTomlParser.java:281]: Check the casts for the following cases as well 
```ballerina
type AuthInfo readonly & record {|
    readonly string username;
    int id = 100;
    ...
|};
```
And this 
```ballerina
type AuthInfo record {|
    readonly string username;
    readonly int id = 100;
    //all fields are readonly
|};
```

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/configurable/ConfigTomlParser.java [hunk 2]
```diff
@@ -228,22 +242,37 @@ private static Object retrieveRecordValues(TomlTableNode tomlValue, String varia
                 default:
                     objectValue = retrievePrimitiveValue(value, variableName + "." + fieldName, fieldType);
             }
-            initialValueEntries.put(fieldName, objectValue);
+            initialValueEntries[count] = new MappingInitialValueEntry.KeyValueEntry(StringUtils.fromString(fieldName)
+                    , objectValue);
+            count++;
         }
-        validateRequiredField(initialValueEntries, effectiveType, variableName);
```
Comment (HindujaB) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/configurable/ConfigTomlParser.java:233]: specialise the field read-only case to use mapValue API

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/configurable/ConfigTomlParser.java [hunk 3]
```diff
@@ -278,8 +288,34 @@ private static Object retrieveRecordValues(TomlNode tomlNode, String variableNam
...
+        return ValueCreator.createReadonlyRecordValue(recordType.getPackage(), recordName, initialValueEntries);
+    }
+
+    private static Object createReadOnlyFieldRecord(Map<String, Object> initialValueEntries, RecordType recordType,
+                                                    String variableName) {
+        for (Map.Entry<String, Field> field : recordType.getFields().entrySet()) {
+            String fieldName = field.getKey();
+            long flag = field.getValue().getFlags();
+            // remove after fixing #28966
+            if (!SymbolFlags.isFlagOn(flag, SymbolFlags.OPTIONAL) && !SymbolFlags.isFlagOn(flag,
```
Comment (HindujaB) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/configurable/ConfigTomlParser.java:303]: If a record with all `readonly` fields contains a defaultable field and the value is not provided at configuration, we throw an error.

File: tests/jballerina-integration-test/src/test/resources/configurables/configurableProject/main/main.bal [hunk 4]
```diff
@@ -38,11 +39,30 @@ type AuthInfo record {|
...
+    readonly string name = "Default";
+    readonly float salary?;
+|};
+
+type Person readonly & record {|
+    readonly string name;
+    string address = "default address";
+    int age?;
+|};
+
```
Comment (manuranga) [inline tests/jballerina-integration-test/src/test/resources/configurables/configurableProject/main/main.bal:53]: Let's add a use case for no-fields-are-read-only record.

**Review by manuranga — CHANGES_REQUESTED** [review]

**Review by manuranga — DISMISSED** [review]

**Review by warunalakshitha — APPROVED** [review]
LGTM

**Pull request merged**